### ISSUE TITLE
User Sources

### ISSUE THREAD

**Issue by bluss** [kupferlauncher/kupfer#78]
Allow users to extend the catalog by using shell scripts (or other programs) that produce textual output. In this way they act like indexers. A User source plugin will parse the output of each command into Leaves of the correct type.

**Comment by SanskritFritz** [comment]
Great idea!

**Comment by khurshid-alam** [comment]
Wonderful idea.

**Comment by bluss** [comment]
Cool. Do you have any wishes for the format?

Supporting a plain list of files (one file per line) is clear candidate. 

What I would reach for first is JSON, in a prototype I have one JSON object (dict) per line. JSON is good because it is typed, and we could use this to have it be entirely extensible (you could create leaves of any loaded plugin that whose leaves support a protocol for creating them from a dict from JSON).

**Comment by khurshid-alam** [comment]
Yes JSON is nice. 

> you could create leaves of any loaded plugin that whose leaves support a protocol for creating them from a dict from JSON

Can you elaborate?

**Comment by bluss** [comment]
Not yet, haven't gotten that far in the design.

**Comment by khurshid-alam** [comment]
Alright.

So it will index anything. Let's say I want to recursively index all png file from ~/Pictures/Screenshot at toplevel. We can do that by presenting all files (or file names) as json obj?

**Comment by bluss** [comment]
Yes, which leaves are created depends entirely on the output of the script. If it outputs one JSON record (if that's the format) per png file, those files are all part of the user source.

**Comment by bluss** [comment]
Json is easy to output from Python, and using external python scripts is not useless, but maybe there are other formats that are better. Big value in using an existing format with existing parser and typed values though.

**Comment by khurshid-alam** [comment]
What about action? We still want to do something with it? For example opening those pngs with default image viewer?

**Comment by bluss** [comment]
I imagine they would show up as files, so they are just like any other .png file in kupfer (so the same actions).

It would be open for a plugin to implement a different leaf for them, but I don't think files should be represented in any other way than FileLeaf in kupfer.

**Comment by bluss** [comment]
So to clarify, the design IMO would be that user sources is a way to insert collections of objects (Leaves in Kupfer's terms) with the help of text based programs, using any of the builtin or plugin defined Leaves that support creation from (for example) JSON. It's not intended that user sources defines new classes of leaves.

**Comment by hugosenari** [comment]
**Nice idea!**

Now we need:
- Define how user will set this sources
  - One or more sources
  - Static file or exec output
  - At preference interface or put exec files at specific directory (like plugins)
- Define expected output from this sources
   - JSON or CSV
   - What information is required
- Define Leaf interface
   - static method or kword at init

**Comment by hugosenari** [comment]
I like JSON but unix commands can easily output a 'CSV' (or CSV like output).

**Comment by bluss** [comment]
If the csv has named columns it should work well, then it's easy to map each row to a dictionary just like JSON; however then we don't have types. Majority will use strings anyway, but I don't want to distribute the lowest level of input parsing to the plugins, it should be central.

**Comment by hugosenari** [comment]
I liked this [repo](https://github.com/funivan/kupfer-custom-urls) is some kind of early stage of what you proposed here.
And thinking about maybe we can adopt [intents](http://stackoverflow.com/questions/2448213/how-to-implement-my-very-own-uri-scheme-on-android) here ie:
- tel:987654321 (is a telefone leaf);
- http://asdfgh (is url leaf);
- mailto://hugosenari@domain.com (is a mail leaf);
- file:///opt/kupfer/kupfer.png (is file leaf)

But, again would be hard create them from command lines. 🤔

**Comment by hugosenari** [comment]
A long time ago I found @spiritedflow [kupfer-connectables](https://github.com/spiritedflow/kupfer-connectables) plugin. For some reason a forgot them. [LOL](https://github.com/spiritedflow/kupfer-connectables/issues/1)

It works with YAMLs 💩:
One for [Leafs](https://github.com/spiritedflow/kupfer-connectables/blob/master/example-config/connectables/leafs/) 
```yaml
locale: en_EN.UTF-8
cacti_id: 1
icon_name: network-server
actions:
  - ssh
  - cacti
  - ping
```

And other for [Actions](https://github.com/spiritedflow/kupfer-connectables/blob/master/example-config/connectables/actions/):
```yaml
type: uri
name: Cacti charts
descr: View cacti charts for this host
icon_name: web-browser
uri: https://your-cacti-site/cacti/graph_view.php?action=tree&tree_id=6&leaf_id=%(cacti_id)d
```

Once again, is a bit hard create this files from command lines output.
But for user configuration is easy.

**Comment by hugosenari** [comment]
And the Great Bedkowski created [user_sources.py](https://github.com/KarolBedkowski/kupfer-contrib/blob/master/plugins/user_sources.py)
